pkg/controller/grafanadashboard/dashboard_pipeline.go

```diff
@@ -109,7 +111,7 @@ func (r *DashboardPipelineImpl) validateJson() error {
 // 3) no configmap specified: try to use embedded json
 // 4) no json specified: try to use embedded jsonnet
 func (r *DashboardPipelineImpl) obtainJson() error {
-	if r.Dashboard.Spec.Url != "" {
+	if r.Dashboard.Spec.Url != "" && len(r.Dashboard.Spec.Json) == 0 {
```

Comment:
I failed to test a case whereby the json isn't empty, as is the case with [DashboardFromURL](https://github.com/integr8ly/grafana-operator/blob/master/deploy/examples/dashboards/DashboardFromURL.yaml), where the JSON is used to notify that the repo is down, I'd suggest to implement a further check here to check if the json from URL and spec JSON are different and then overwrite the spec with the fetched JSON, thanks!

Reply:
>I'd suggest to replace the len(r.Dashboard.Spec.Json) == 0 (or add another condition) to check if the fetched URL-JSON is different to the Spec-JSON.

JSON would take precedence in this change. There is no way to know if the URL has changed it contents, so we'd have to refetch it constantly, so well hit HTTP 409 eventually.

Luckily grafana.com dashboards have revisions, so URL contents doesn't change. Should I update example dashboard to point to a particular commit?

Reply:
I see your point, and I agree that it's not ideal to refetch it constantly (it doesn't fix the problem at all)
>Luckily grafana.com dashboards have revisions, so URL contents doesn't change. Should I update example dashboard to point to a particular commit?

this is only really getting around the problem with verifying it with this specific example dashboard but there's another (bigger)problem and that would be to account for backward compatibility, we would like to avoid manual intervention in upgrades wherever possible, this would mean that the way it's implemented here, it would break existing grafanadashboard deployments where on upgrade if `url != "" ` and `spec.json != ""` then the URL is never fetched and the json is constantly taking precedence even if it holds invalid (or vastly out-dated) data such as the case with the example dashboard, thus requiring manual deletion of the pre-defined JSON, It's been a bit of a long day today and my imagination isn't quite serving me right now to get an idea on how to get around this. I'll try and figure a way out tomorrow.
But It might be the case that we just have to push a breaking change and do a major version bump if we want this bug fixed 🤔 , Peter might have a better idea when he gets some time to look at this.

Reply:
Having thought about this, another work around might be possible if we extend  the periodical reconcile, previously there was a problem with dashboards not being discovered until that time has elapsed, but it might be worthwhile to look into resolving that now, This could maybe also resolve another issue that someone had brought up in the slack channel (about too many requests being sent).
Extending the reconcile loop should fix  `429 Too Many Requests` as we would then be hitting that endpoint a lot less

Reply:
>it would break existing grafanadashboard deployments where on upgrade if url != "" and spec.json != "" 

Correct, perhaps its worth including in the changelog. I don't think `url != "" && json != ""` is common though - I can't imagine the usecase where both are set (json contents would be ignored, so why set it?), so I wouldn't worry too much about breaking existing installs.

Its certainly worth including in the changelog, as it might be counter intuitive.

Reply:
Another option is to have a `.status.content-length` where the size of fetched json (no matter if from `.spec.url` or `.spec.json`) is stored. This would allow us to determine if the URL has changes via HEAD request headers. This method would work with grafana.com, but may not work on other servers

Reply:
The original idea was the you set both, `url` and `json` so that when the URL is down, a fallback dashboard can be shown. Currently the order of precedence is url -> configmap -> json -> jsonnet. I'm doubtful that this is really useful for the following reasons:

1. the best fallback dashboard is the last one fetched from the URL
2. the user can still provide a `json` value and it will be used if the first request already fails

 I like the idea with the `HEAD` request also but it still involves periodic network requests.